File: ch02/src/relaxed_ordering_stop_flag.cpp
#include <atomic>
#include <iostream>
#include <string>
#include <thread>

using namespace std::chrono_literals;

void some_work() noexcept
{
    std::this_thread::sleep_for(1s);
}

int main()
{
    auto stop = std::atomic_bool { false };

    auto background_thread = std::jthread {
        [&]() {
            while (!stop.load(std::memory_order_relaxed)) {
                some_work();
            }
        }
    };

    std::string line {};
    while (true) {
        std::getline(std::cin, line);
        if (line == "help") {
            std::cout << "Commands: help, stop\n";
        } else if (line == "stop") {
            stop.store(true, std::memory_order_relaxed);
            break;
        } else {
            std::cout << "Unknown command: " << line << '\n';
        }
    }

    return 0;
}

File: ch02/src/bin/relaxed_ordering_stop_flag.rs
use std::{
    sync::atomic::{AtomicBool, Ordering},
    thread,
    time::Duration,
};

/// Usage example of a stop flag with relaxed memory ordering.
/// This ordering is OK because we're working on one (1) flag.
fn main() {
    static STOP: AtomicBool = AtomicBool::new(false);

    // Working in the background
    let background_thread = thread::spawn(|| {
        while !STOP.load(Ordering::Relaxed) {
            thread::sleep(Duration::from_millis(10));
        }
    });

    // Listen to user input in the main thread
    for line in std::io::stdin().lines() {
        match line.unwrap().as_str() {
            "help" => println!("commands: help, stop"),
            "stop" => break,
            cmd => println!("unknown command: {cmd}"),
        }
    }

    // Request the background thread to stop
    STOP.store(true, Ordering::Relaxed);

    background_thread.join().unwrap();
}
